### PR TITLE
RFC 004: Dart implementation

### PR DESCRIPTION
**Pull request by srikrsna-buf**
Add an RFC for creating Connect-Dart.

### CODE REVIEW

**Review by timostamm — COMMENTED** [review]
Would be great to have support for Dart clients 👍 

Since the implementation depends on the platform's HTTP stack, I wonder which platforms exactly can be supported (and continually tested via the conformance suite). 

Related, I wonder who will be maintaining the implementation?

File: docs/governance/rfc/004-dart-implementation.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+- [`HttpClient`][dart-io-client] from `dart:io`. This is part of the std library and runs on native platforms.
+  This only supports HTTP/1.
+- [`package:http2`][pub-http2]. This is a HTTP/2 implementation maintained by the Dart team. This can be used
+  to support gRPC. Connection management has to be implemented on top of this.
+- [`dart:js_interop`][dart-js-interop]: This is part of the std library and can be used to create
+  bindings for web APIs. We can use the `fetch` APIs to support Connect and gRPC-Web.
+
+Platforms like Android and iOS have their own HTTP stack, and it is desirable to use them. [`package:http`][pub-http] 
+can be used as a cross platform client to support both. It is maintained by the Dart team. The only blocker
+is the missing support for [aborting requests][http-abort-issue].
```
Comment (jhump) [inline docs/governance/rfc/004-dart-implementation.md:30]: Will this be an actual _blocker_? If so, what will use instead for an initial release? Or will we use this but it blocks promoting to v1.0 until we can support aborting requests? Maybe clarify a little how this issue impacts how we intend to implement support for Android and iOS.

**Review by timostamm — APPROVED** [review]

**Review by jhump — COMMENTED** [review]
LGTM

File: docs/governance/rfc/004-dart-implementation.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+Dart is a popular programming language powering frameworks like Flutter which
+is used for cross-platform app development. This will help grow Connect's
+adoption.
+
+## Anticipated complications
+
+Dart runs on several platforms: Web, Android, and iOS to name a few. Most of the platforms
+have their own HTTP stack with varying levels of support for HTTP specs. It doesn't have a
+unified abstration and APIs for common HTTP constructs like headers and cancellation.
```
Comment (smaye81) [inline docs/governance/rfc/004-dart-implementation.md:16]:
```suggestion
unified abstraction and APIs for common HTTP constructs like headers and cancellation.
```

File: docs/governance/rfc/004-dart-implementation.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+## Supported features
+
+Clients should support all three protocols and four RPC kinds where possible:
+
+- Connect, gRPC-Web:
+  - Should be supported on all platforms.
+  - On the web, unary and server stream can be supported.
+- gRPC:
+  - Can be supported on native playforms with support for HTTP/2
```
Comment (smaye81) [inline docs/governance/rfc/004-dart-implementation.md:42]:
```suggestion
  - Can be supported on native platforms with support for HTTP/2
```

File: docs/governance/rfc/004-dart-implementation.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+These should be continously tested using the [conformance suite][conformance] on both native and web platforms.
```
Comment (smaye81) [inline docs/governance/rfc/004-dart-implementation.md:44]:
```suggestion
These should be continuously tested using the [conformance suite][conformance] on both native and web platforms.
```

**Review by smaye81 — APPROVED** [review]

**Review by jhump — APPROVED** [review]

**Review by emcfarlane — APPROVED** [review]

File: docs/governance/rfc/004-dart-implementation.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+have their own HTTP stack with varying levels of support for HTTP specs. It doesn't have a
+unified abstraction and APIs for common HTTP constructs like headers and cancellation.
+
+We have to create a minimal HTTP abstraction to support various platforms. We can then have concrete
+implementations using the following:
+
+- [`HttpClient`][dart-io-client] from `dart:io`. This is part of the std library and runs on native platforms.
+  This only supports HTTP/1.
+- [`package:http2`][pub-http2]. This is a HTTP/2 implementation maintained by the Dart team. This can be used
+  to support gRPC. Connection management has to be implemented on top of this.
```
Comment (pkwarren) [inline docs/governance/rfc/004-dart-implementation.md:24]:
```suggestion
  to support bidirectional streaming and gRPC. Connection management has to be implemented on top of this.
```

**Review by pkwarren — APPROVED** [review]

**Review by akshayjshah — APPROVED** [review]

**Pull request merged**